synapse/handlers/pagination.py

```diff
@@ -381,22 +379,28 @@ async def get_messages(
                         member_event_id
                     )
                     if RoomStreamToken.parse(leave_token).topological < max_topo:
-                        source_config.from_key = str(leave_token)
+                        from_token = from_token.copy_and_replace(
+                            "room_key", leave_token
+                        )
 
                 await self.hs.get_handlers().federation_handler.maybe_backfill(
                     room_id, max_topo
                 )
 
+            to_room_key = None
+            if pagin_config.to_token:
+                to_room_key = pagin_config.to_token.room_key
```

Comment:
I'm not sure where this matches up with the old version of this code? It seems previously we just always used what was in `source_config`?

Reply:
There was special logic in `get_source_config` to handle the fact that `pagin_config.to_token` could be `None`, which we need to hoist up here. Bear in mind that `source_config.to_key` is the room token rather than the stream token (i.e. `source_config.to_key = pagin_config.to_token.room_key if pagin_config.to_token else None`)